airbyte-integrations/connectors/destination-clickhouse-v2/src/main/kotlin/io/airbyte/integrations/destination/clickhouse_v2/write/load/BinaryRowInsertBuffer.kt

```diff
@@ -37,20 +53,48 @@ private val log = KotlinLogging.logger {}
  */
 @SuppressFBWarnings(
     value = ["NP_NONNULL_PARAM_VIOLATION"],
-    justification = "suspend and fb's non-null analysis don't play well"
+    justification = "suspend and fb's non-null analysis don't play well",
 )
 class BinaryRowInsertBuffer(
     val tableName: TableName,
     private val clickhouseClient: Client,
+    private val airbyteSchema: AirbyteType
 ) {
     // Initialize the inner buffer
-    private val schema = clickhouseClient.getTableSchema(tableName.name, tableName.namespace)
+    private val schema: TableSchema =
+        clickhouseClient.getTableSchema(tableName.name, tableName.namespace)
     @VisibleForTesting internal var inner = InputOutputBuffer()
     @VisibleForTesting
     internal var writer = RowBinaryFormatWriter(inner, schema, ClickHouseFormat.RowBinary)
 
+    internal val formattedSchema =
+        if (airbyteSchema is ObjectType) {
+            airbyteSchema.asColumns().mapKeys { (k, _) -> k.toClickHouseCompatibleName() }
+        } else {
+            emptyMap()
+        }
+
     fun accumulate(recordFields: Map<String, AirbyteValue>) {
-        recordFields.forEach { writeAirbyteValue(it.key, it.value) }
+        recordFields.forEach {
+            val airbyteType: AirbyteType =
+                if (COLUMN_NAMES.contains(it.key)) {
+                    when (it.key) {
+                        Meta.COLUMN_NAME_AB_RAW_ID -> StringType
+                        Meta.COLUMN_NAME_AB_META -> StringType
+                        Meta.COLUMN_NAME_AB_EXTRACTED_AT -> DateType
+                        Meta.COLUMN_NAME_AB_GENERATION_ID -> IntegerType
+                        else -> // Unreachable
+                        StringType
+                    }
+                } else {
+                    if (airbyteSchema is ObjectType) {
+                        formattedSchema[it.key]!!.type
+                    } else {
+                        airbyteSchema
```

Comment:
not sure I understand what this is doing. Why are we applying the root-level schema to each individual value?

Reply:
I have updated it. We need to fetch it from the formattedSchema map to get the clickhouse compatible column names